### PR TITLE
fix(glm): prepare B12X startup contracts for target and draft models

### PR DESCRIPTION
**Pull request by voipmonitor**
## Behavior

Prepare the B12X operations that GLM target and draft models actually execute before CUDA graph capture:

- Register GLM mHC after broadcast weights are published, with explicit model-specific norm/projection operands. Retain prepared row plans across workload stages. DeepSeek's BF16 FFN operand remains supported.
- Declare per-dimension KDA forget gates and preserve the fused beta view's stride and alignment during preparation and tuning.
- Declare all 32 replicated indexer heads and the bound pooled-page table capacity. Give explicitly named and anonymous indexers distinct preparation identities.
- Bound preparation keys/lengths by the page table's addressable capacity, including narrow DCP-local tables. Keep the backing-cache capacity and top-k tuning bounds as additional limits.
- Register MTP vocabulary projections, share plans for an aliased target/draft head without creating an `nn.Module` cycle, and exclude heads that already own NVFP4/MXFP8 preparation.
- Collect target, draft and communication plans from the combined row-count workload, with separate execution lanes.

No environment switch or attention arithmetic change is introduced. Native serialized FP8 checkpoints keep their BF16 KDA projections; ModelOpt's per-projection quantization is preserved.

## Relationship to #761

This is a consolidated alternative to #761, **not an additional PR to merge after it**. It combines the independently serving-qualified NVFP4 startup fixes with Toby Mao's operand declarations, plan retention, trial-layout and shared-head/workload fixes. Toby is credited as co-author and #761 remains the source reference.

Material differences:

- Keep the model-wide quantization configuration unchanged; only the native serialized FP8 adapter receives a shallow copy with KDA quantization disabled. #761 clears every quantization configuration, including ModelOpt. A four-case constructor regression passes here; the #761 constructor fails the three non-`None` configuration invariants.
- Derive beta's row width from the actual fused projection, including padding, rather than reconstructing it from head counts.
- Preserve active-width prefill page-table slicing. B12X already prepares the short-width execution variant; forcing every request to process maximum-width metadata is unnecessary.
- Preserve JJ's cache-view strides, cache-identity invalidation and meaningful indexer tuning inputs. Retain unique fallback names when a caller supplies no prefix.
- Keep JJ's existing FP8 preparation-key correction; do not replace it with the older `plans` fix from #761.

KDA prefill accepts beta strides as runtime integers; it does not require a full fused-projection-sized persistent beta allocation. That path retains its compact staging allocation. Decode's compile-time layout is reproduced exactly.

## Composition

For the qualified NVFP4 serving stack, combine with #769, #770, #771 and local-inference-lab/b12x#379. #771 carries sparse-MLA discovery, shared scratch, physical-selection preparation and result normalization. Its naming overlap is moved here, so the branches merge without manual conflict resolution. Two-shot is disabled in serving measurements; local-inference-lab/b12x#380 is independently qualified.

## Validation

Status: implemented; GLM NVFP4 TP4/DCP1 DFlash2 K7 serving-qualified on 4× RTX PRO 6000 Workstation. Native FP8/GB10 end-to-end serving is not qualified by these tests.

- 32 focused startup regressions pass, including GLM/DeepSeek mHC, padded gate layouts, plan retention, vocabulary ownership, FP8 keys, combined workloads and narrow bound page tables. Page widths 1/4 fail before the capacity correction and pass afterward; width 32 preserves the same 1024-key fixture.
- Full GLM model, pooled-indexer and B12X warmup suites: **166 passed, 41 skipped**. Optional tests explain their skips; this includes GPU KDA/reference and graph tests. Counts overlap the focused suite.
- All changed-file pre-commit checks pass, including mypy.
- Composed Docker completes target FULL_AND_PIECEWISE and DFlash FULL capture and all measured requests normally.
- MTP3 also completes startup and full graph capture with its NVFP4 draft-only vocabulary head, followed by a warmup and three complete Sieve requests. Median output is 322.92 tok/s and verifier rate 95.65 steps/s at temperature1/top-p.95. This is a functional smoke test, not matched MTP performance qualification; first-use metadata JIT warnings remain outside the prepared B12X operations.

Same GPUs 4–7, VRAM +6000, TP4/DCP1, batch4096, temperature1/top-p.95, W8A8 draft, B12X PCIe AR, two-shot off:

| Measurement | Serving reference | This composition | Change |
| --- | ---: | ---: | ---: |
| Sieve C1 output, median of 15 requests | 552.76 tok/s | 547.85 tok/s | −0.89% |
| Verifier steps, median of 15 requests | 92.91/s | 92.81/s | −0.11% |
| Accepted tokens per step, median | 4.915 | 4.855 | −1.23% |
| Uncached 32k prefill, 30-second window | 15,880 tok/s | 15,856 tok/s | −0.15% |

Both prefill arms contain 12 requests with 32,315 input tokens. The metric is client input tokens/TTFT, not isolated GPU prefill time. Sieve uses identical seeds and a complete warmup, but speculative output is stochastic. These bounded checks support performance preservation, not a speedup or a general quality guarantee. Available KV memory is 27.59 GiB/GPU.

Test commands, in a matching CUDA/B12X environment with repository test dependencies:

```sh
python -m pytest -q tests/models/test_glm5next_model.py \
  tests/models/test_glm5next_pooled_indexer.py \
  tests/model_executor/test_b12x_warmup.py
python -m pytest --noconftest -q tests/model_executor/kernels/test_b12x_linear.py \
  -k 'mhc_declares or mhc_keeps or fp8_preparation_units_key'
```

Matched measurement source: JJ `e0f22d9c598`, startup branch `e7fc5e21330`, clean PR composition `835c8dcca87`; B12X `5c50b153386` plus [B12X #379](https://github.com/local-inference-lab/b12x/pull/379) and [B12X #380](https://github.com/local-inference-lab/b12x/pull/380). Startup head `6276a0bdde8` adds the reproduced narrow-page preparation bound; its clean composition is `2ba5dcee206`. The bound leaves the measured 1M DCP1 tuning fixture unchanged because its table addresses more than 1024 keys. The rebuilt composition also passes DFlash startup, graph capture and two complete Sieve requests, including warmup. Diagnostic Docker builds change Python sources only over the serving reference; they are not registry releases or claims that the native runtime was rebuilt.

AI-assisted implementation and validation; maintainer review requested.

### CODE REVIEW

**Comment by voipmonitor** [comment]
@coderabbitai review

Please review the consolidated GLM startup contracts, particularly quantization-config isolation, target/draft plan ownership and descriptor/runtime layout consistency. The PR describes bounded DFlash serving qualification and the unqualified native FP8/GB10 scope.

File: vllm/models/deepseek_v4/nvidia/b12x_indexer.py [hunk 1]
```diff
@@ -446,13 +471,18 @@ def make_call(state, *, caps=caps, mode=mode, snapshot=snapshot):
                         device=caps.device,
                     )
                     pages = torch.zeros(
-                        (caps.max_q_rows, caps.max_page_table_width),
+                        (
+                            1 if mode == "prefill" else caps.max_q_rows,
+                            caps.max_page_table_width,
+                        ),
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/models/deepseek_v4/nvidia/b12x_indexer.py:478]: Confirmed with a minimal prepare-call reproducer: a 1024-key cache and bound table widths 1 or 4 failed with tensor-size errors; width 32 passed. Commit 5fd21e1737a caps live_keys at caps.max_page_table_width * 64, preserving matching lengths, active width and page IDs. All three cases now pass for both decode and prefill, including shared-page stride and cache restoration. The 1M DCP1 measured fixture remains unchanged because its table capacity exceeds the existing 1024-key tuning bound. The fix is in this PR; the focused startup suite passes 32 cases.
